[Datasets] Add documentation for dynamic block splitting (#30480)
ray-project/ray: doc/source/data/dataset-internals.rst

Execution Memory
~~~~~~~~~~~~~~~~

> During execution, a task can read multiple input blocks, and write multiple output blocks. The input blocks consume object stored shared memory (and Python heap memory if conversion to non-Arrow format is needed). The output blocks consume Python heap memory (prior to putting in the object store) as well as object store memory (after being put in the object store).

Reviewer: ```suggestion
During execution, a task can read multiple input blocks, and write multiple output blocks. Input and output blocks consume both worker heap memory and shared memory via Ray's object store.
```

Reviewer: This is simpler and I don't think the other details are necessary.

Author: @stephanie-wang - thanks, changed.

Datasets attempts to bound its heap memory usage to `num_execution_slots * max_block_size`. The number of execution slots is by default equal to the number of CPUs, unless custom resources are specified. The maximum block size is set by the configuration parameter `ray.data.context.DatasetContext.target_max_block_size` and is set to 512MiB by default. When a task's output is larger than this value, the worker will automatically split the output into multiple smaller blocks to avoid running out of heap memory.

> Large block size can lead to potential out-of-memory situations. To avoid these issues, make sure no single item in your Datasets is too large, and always call :meth:`ds.map_batches() <ray.data.Dataset.map_batches>` with batch size small enough such that the output batch can comfortably fit into memory.

Author: Still keep the original statements here, guess it might be better than nothing. Will remove it once adding the OOM section in troubleshooting guide.

> Certain types of intermediate data must fit in memory. When a task has multiple input blocks, only one input block needs to fit in memory at any given time. When a task has multiple output blocks, before Ray 2.2 all output blocks need to fit in memory. After Ray 2.2 with feature of "dynamic blocks splitting", only one output block needs to fit in memory at any given time. Under the hood, Ray Datasets uses :ref:`remote generators <generators>` to put output blocks in the object store one by one, without holding all blocks in memory. Dynamic blocks splitting is enabled by default after Ray 2.2. It supports Ray task, and is not yet supported for Ray actor and Ray client. It support one-to-one :ref:`transformations <transform_datasets_transformations>` (such as ``ray.data.read_*()`` and :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`), and is not yet supported for all-to-all transformations (such as :meth:`ds.random_shuffle() <ray.data.Dataset.random_shuffle>`).

Reviewer 2: nit: I'm not sure what best practices are around mentioning version differences, but given that we have versioned docs, would it make sense to just have this describe the current state of the world (i.e. Ray 2.2)?

Author: @matthewdeng - changed the working, removed the Ray version.

> Certain types of intermediate data must fit in memory. When a task has multiple input blocks, only one input block needs to fit in memory at any given time. When a task has multiple output blocks, before Ray 2.2 all output blocks need to fit in memory. After Ray 2.2 with feature of "dynamic blocks splitting", only one output block needs to fit in memory at any given time. Under the hood, Ray Datasets uses :ref:`remote generators <generators>` to put output blocks in the object store one by one, without holding all blocks in memory. Dynamic blocks splitting is enabled by default after Ray 2.2. It supports Ray task, and is not yet supported for Ray actor and Ray client. It support one-to-one :ref:`transformations <transform_datasets_transformations>` (such as ``ray.data.read_*()`` and :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`), and is not yet supported for all-to-all transformations (such as :meth:`ds.random_shuffle() <ray.data.Dataset.random_shuffle>`).

Reviewer 2: When would Ray actors be used? One suggestion is to move all the cases where it's not working into a consolidated `Warning` section to improve readability.

Author: @matthewdeng - changed the wording, removed the limitation for now. btw Ray actor can be used in `map_batches` with actor pool strategy.

> Certain types of intermediate data must fit in memory. When a task has multiple input blocks, only one input block needs to fit in memory at any given time. When a task has multiple output blocks, before Ray 2.2 all output blocks need to fit in memory. After Ray 2.2 with feature of "dynamic blocks splitting", only one output block needs to fit in memory at any given time. Under the hood, Ray Datasets uses :ref:`remote generators <generators>` to put output blocks in the object store one by one, without holding all blocks in memory. Dynamic blocks splitting is enabled by default after Ray 2.2. It supports Ray task, and is not yet supported for Ray actor and Ray client. It support one-to-one :ref:`transformations <transform_datasets_transformations>` (such as ``ray.data.read_*()`` and :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`), and is not yet supported for all-to-all transformations (such as :meth:`ds.random_shuffle() <ray.data.Dataset.random_shuffle>`).

Reviewer 3: Is this really true? If the task wants to perform an operation across all its input blocks, they need to present in memory.

Author: @jianoaix - yes only when user sets a very large batch_size, which exceed total size of all blocks. Changed the wording, here.

> Certain types of intermediate data must fit in memory. When a task has multiple input blocks, only one input block needs to fit in memory at any given time. When a task has multiple output blocks, before Ray 2.2 all output blocks need to fit in memory. After Ray 2.2 with feature of "dynamic blocks splitting", only one output block needs to fit in memory at any given time. Under the hood, Ray Datasets uses :ref:`remote generators <generators>` to put output blocks in the object store one by one, without holding all blocks in memory. Dynamic blocks splitting is enabled by default after Ray 2.2. It supports Ray task, and is not yet supported for Ray actor and Ray client. It support one-to-one :ref:`transformations <transform_datasets_transformations>` (such as ``ray.data.read_*()`` and :meth:`ds.map_batches() <ray.data.Dataset.map_batches>`), and is not yet supported for all-to-all transformations (such as :meth:`ds.random_shuffle() <ray.data.Dataset.random_shuffle>`).

Reviewer: ```suggestion
Datasets attempts to bound its heap memory usage to `num execution slots * max block size`.
The number of execution slots is by default equal to the number of CPUs, unless custom resources are specified.
The maximum block size is set by the configuration parameter `ray.data.context.DatasetContext.target_max_block_size` and is set to 512MiB by default.
When a task's output is larger than this value, the worker will automatically split the output into multiple smaller blocks to avoid running out of heap memory.

See the troubleshooting guide for more details on handling out-of-memory errors.
```

Reviewer: Simplified to give only relevant details. Agree with @matthewdeng that it's probably fine not to list all the limitations here as long as we give good error messages at run time.

I would also remove the paragraph below and move the content to the common troubleshooting guide.
